File: cms-server/middleware/audit.js
/**
 * 操作日志中间件
 * 在需要记录的操作路由里手动调用 audit(log), 或由中间件自动记录
 */

const Database = require('better-sqlite3');
const path = require('path');

const DB_PATH = path.join(__dirname, '../db/cms.db');

function getDB() {
  return new Database(DB_PATH);
}

/**
 * 写入审计日志
 * @param {object} req   - express request
 * @param {string} action - 操作类型，如 'login' / 'update_page' / 'create_user'
 * @param {string} target - 操作目标，如 'visa' / 'user:3'
 * @param {string} [detail] - 补充说明
 */
function audit(req, action, target, detail) {
  const db = getDB();
  try {
    db.prepare(
      `INSERT INTO audit_log (user_id, username, action, target, detail)
       VALUES (?, ?, ?, ?, ?)`
    ).run(
      req.user?.id || null,
      req.user?.username || 'anonymous',
      action,
      target || '',
      detail || ''
    );
  } catch (e) {
    console.error('[audit] 写入失败:', e.message);
  } finally {
    db.close();
  }
}

/**
 * Express 中间件：自动记录所有写操作（POST/PUT/DELETE）
 * 放在路由之前，只记录不阻断
 */
function auditMiddleware(req, res, next) {
  const originalSend = res.send;
  // 拦截响应，只在成功时记录
  res.send = function (body) {
    if (req.method !== 'GET' && res.statusCode < 400) {
      // 异步写日志，不阻塞响应
      setImmediate(() => {
        try {
          const db = getDB();
          db.prepare(
            `INSERT INTO audit_log (user_id, username, action, target, detail)
             VALUES (?, ?, ?, ?, ?)`
          ).run(
            req.user?.id || null,
            req.user?.username || 'anonymous',
            `api_${req.method.toLowerCase()}`,
            req.route?.path || req.path,
            JSON.stringify(req.body).slice(0, 200)
          );
          db.close();
        } catch {}
      });
    }
    return originalSend.call(this, body);
  };
  next();
}

module.exports = { audit, auditMiddleware, getDB };

File: cms-server-go/middleware/audit.go
package middleware

import (
	"cms-server-go/db"
	"io"
	"log"
	"strings"

	"github.com/gin-gonic/gin"
)

// Audit 写入审计日志
func Audit(c *gin.Context, action, target, detail string) {
	userID, _ := c.Get("userID")
	username, _ := c.Get("username")

	var uid *int64
	if id, ok := userID.(int64); ok && id > 0 {
		uid = &id
	}

	uname := "anonymous"
	if u, ok := username.(string); ok && u != "" {
		uname = u
	}

	_, err := db.DB.Exec(
		"INSERT INTO audit_log (user_id, username, action, target, detail) VALUES (?, ?, ?, ?, ?)",
		uid, uname, action, target, detail,
	)
	if err != nil {
		log.Printf("[audit] 写入失败: %v", err)
	}
}

// AuditMiddleware 自动记录所有写操作（POST/PUT/DELETE）
func AuditMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Next()

		// 只在写操作且响应成功时记录
		method := c.Request.Method
		if method == "GET" {
			return
		}
		if c.Writer.Status() >= 400 {
			return
		}

		userID, _ := c.Get("userID")
		username, _ := c.Get("username")

		var uid *int64
		if id, ok := userID.(int64); ok && id > 0 {
			uid = &id
		}

		uname := "anonymous"
		if u, ok := username.(string); ok && u != "" {
			uname = u
		}

		action := "api_" + strings.ToLower(method)
		target := c.FullPath()
		if target == "" {
			target = c.Request.URL.Path
		}

		// 限制 detail 长度
		detail := ""
		if c.Request.Body != nil {
			bodyBytes, _ := io.ReadAll(c.Request.Body)
			if len(bodyBytes) > 200 {
				detail = string(bodyBytes[:200])
			} else {
				detail = string(bodyBytes)
			}
		}

		go func() {
			_, _ = db.DB.Exec(
				"INSERT INTO audit_log (user_id, username, action, target, detail) VALUES (?, ?, ?, ?, ?)",
				uid, uname, action, target, detail,
			)
		}()
	}
}
